cogs/commands/quotes.py

```diff
@@ -0,0 +1,450 @@
+import logging
+import re
+from datetime import datetime
+from enum import Enum, auto, unique
+from functools import singledispatch
+from typing import Optional, Union
+
+from discord import AllowedMentions
+from discord.ext import commands
+from discord.ext.commands import Bot, Context, Converter
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.sql import func
+from sqlalchemy_utils import ScalarListException
+
+from models import db_session
+from models.quote import Quote, QuoteOptouts
+from utils import (
+    get_database_user,
+    get_name_string,
+    is_compsoc_exec_in_guild,
+    user_is_irc_bot,
+)
+from utils.mentions import Mention, MentionConverter, MentionType
+
+LONG_HELP_TEXT = """
+Pull a random quote. Pull quotes by ID using "#ID", by author using "@username", or by topic by entering plain text
+"""
+SHORT_HELP_TEXT = """Record and manage quotes attributed to authors"""
+
+MC = MentionConverter()
+
+
+@unique
+class QuoteError(Enum):
+    BAD_FORMAT = auto()
+    NOT_PERMITTED = auto()
+    NOT_FOUND = auto()
+    OPTED_OUT = auto()
+    DB_ERROR = auto()
+    NO_OP = auto()
+
+
+class QuoteException(Exception):
+    def __init__(
+        self,
+        err: QuoteError,
+        msg=None,
+    ):
+        self.message = msg
+        self.err = err
+
+
+QuoteID = int
+
+
+class QuoteIDConverter(Converter):
+    async def convert(self, ctx, string):
+        if not re.fullmatch("#\d+", string):
+            raise QuoteException(QuoteError.BAD_FORMAT)
+
+        quote_id: QuoteID = int(string[1:])
+        return int(string[1:])
+
+
+def is_id(string: str) -> bool:
+    return re.fullmatch("#\d+", string) is not None
+
+
+def user_opted_out(user: Mention, db_session=db_session) -> bool:
```

Comment:
Could you rename all of these `db_session` parameters to something (e.g. `session`) to make it clearer that they're not the global `db_session` object, please?
```py
def user_opted_out(user: Mention, session=db_session) -> bool:
```